prime-dotnet-webapi/Services/DefaultAutomaticAdjudicationService.cs

```diff
@@ -0,0 +1,213 @@
+                {
+                    // get the current status record for the enrolment
+                    var currentStatus = enrolment.CurrentStatus;
+
+                    // make sure there is a current status
+                    if (currentStatus == null)
+                    {
+                        throw new InvalidOperationException($"Could not process enrolment rule, current status was missing for Enrolment.Id={enrolment.Id}.");
+                    }
+
+                    // for every item returned in the results, add the reason to the current status
+                    foreach (var item in results)
+                    {
+                        if (currentStatus.EnrolmentStatusReasons == null)
+                        {
+                            currentStatus.EnrolmentStatusReasons = new List<EnrolmentStatusReason>(0);
+                        }
+                        currentStatus.EnrolmentStatusReasons.Add(new EnrolmentStatusReason { EnrolmentStatus = currentStatus, StatusReasonCode = item.Code });
+                    }
+                }
+
+                return !ruleFailed;
+            }
+
+            public abstract ICollection<StatusReason> ProcessRuleInternal(Enrolment enrolment);
+
+        }
+
+        /// check to see if any of the self-declaration rules were answered as 'Yes'
+        private class SelfDeclarationRule : BaseAutomaticAdjudicationRule
+        {
+            public override ICollection<StatusReason> ProcessRuleInternal(Enrolment enrolment)
+            {
+                var result = new List<StatusReason>(0);
+                // check to see if any of the self-declaration rules were answered as 'Yes'
+                // note: if for some reason the question was not answered, we will assume 'Yes'
+                if (enrolment.HasConviction.GetValueOrDefault(true)
+                        || enrolment.HasDisciplinaryAction.GetValueOrDefault(true)
+                        || enrolment.HasPharmaNetSuspended.GetValueOrDefault(true)
+                        || enrolment.HasRegistrationSuspended.GetValueOrDefault(true))
+                {
+                    result.Add(new StatusReason { Code = StatusReason.SELF_DECLARATION_CODE });
+                }
+
+                return result;
+            }
+        }
+
+        /// check to see if any of the addresses are outside of BC
+        private class AddressRule : BaseAutomaticAdjudicationRule
+        {
+            public override ICollection<StatusReason> ProcessRuleInternal(Enrolment enrolment)
+            {
+                var result = new List<StatusReason>(0);
+                // check to see if any of the addresses are outside of BC
+                // if there is not code (or address), use BC, so we don't get a mis-match
+                if (string.Compare(Province.BRITISH_COLUMBIA_CODE, (enrolment.Enrollee?.PhysicalAddress.ProvinceCode ?? Province.BRITISH_COLUMBIA_CODE), StringComparison.OrdinalIgnoreCase) != 0
```

Comment:
Perhaps for clarity it might be better to do something like:
```
var provinceCodes = new[] { enrolment.Enrollee.PhysicalAddress?.ProvinceCode, enrolment.Enrollee?.MailingAddress?.ProvinceCode };
                if (provinceCodes.Any(p => p != null && !p.Equals(Province.BRITISH_COLUMBIA_CODE, StringComparison.InvariantCultureIgnoreCase)))
                {
                    result.Add(new StatusReason { Code = StatusReason.ADDRESS_CODE });
                }
```

Reply:
Yes, this makes sense - this check it a little funky, since the data may not be present all the time.